src/JSInterop/Microsoft.JSInterop/src/DotNetObjectRefOfT.cs

```diff
@@ -0,0 +1,117 @@
+    /// but instead should be passed as a reference.
+    ///
+    /// To avoid leaking memory, the reference must later be disposed by JS code or by .NET code.
+    /// </summary>
+    /// <typeparam name="TValue">The type of the value to wrap.</typeparam>
+    public sealed class DotNetObjectRef<TValue> : IDotNetObjectRef, IDisposable where TValue : class
+    {
+        private long? _trackingId;
+
+        /// <summary>
+        /// This A is for meant for JSON deserialization and should not be used by user code.
+        /// </summary>
+        [Obsolete("This API is meant for JSInterop infrastructure and should not be used by user code.")]
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public DotNetObjectRef()
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="DotNetObjectRef{TValue}" />.
+        /// </summary>
+        /// <param name="value">The value to pass by reference.</param>
+        public DotNetObjectRef(TValue value)
+        {
+            Value = value;
+            _trackingId = DotNetObjectRefManager.Current.TrackObject(this);
+        }
+
+        /// <summary>
+        /// Gets the object instance represented by this wrapper.
+        /// </summary>
+        [JsonIgnore]
+        public TValue Value
+        {
+            get;
+            // Workaround for https://github.com/dotnet/corefx/issues/37536 and https://github.com/dotnet/corefx/issues/37567
+            // Once fixed, we should make this property private.
+            set;
+        }
+
+        internal bool Disposed { get; private set; }
+
+        /// <summary>
+        /// This constructor is for meant for JSON serialization and should not be used by user code.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        [JsonPropertyName(DotNetDispatcher.DotNetObjectRefKey)]
+        [Obsolete("This API is meant for JSInterop infrastructure and should not be used by user code.")]
+        public long TrackingId
+        {
+            get
+            {
+                ThrowIfDisposed();
+                return _trackingId.Value;
+            }
+            set
+            {
+                ThrowIfDisposed();
+                if (_trackingId != null)
+                {
+                    throw new InvalidOperationException($"{nameof(DotNetObjectRef<TValue>)} cannot be reinitialized.");
+                }
+
+                _trackingId = value;
+                Value = (TValue)DotNetObjectRefManager.Current.FindDotNetObject(value);
+            }
+        }
+
+        object IDotNetObjectRef.Value => Value;
+
+        /// <summary>
+        /// Implictly converts the <see cref="DotNetObjectRef{TValue}" /> instance to the wrapping value.
+        /// </summary>
+        /// <param name="result">The <see cref="DotNetObjectRef{TValue}"/>.</param>
+        public static implicit operator TValue(DotNetObjectRef<TValue> result) => result.Value;
```

Comment:
I have mixed feelings about this. I appreciate the effort to simplify usage, but if people are using the value as if it is an instance of `TValue`, they might forget there's a wrapper at all, and hence forget to dispose it.

Reply:
Okies. I'll take it out